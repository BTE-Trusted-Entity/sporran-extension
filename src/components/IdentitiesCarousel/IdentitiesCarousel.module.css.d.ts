export const __esModule: true;
export const container: string;
export const chevron: string;
export const left: string;
export const right: string;
export const bubbles: string;
export const item: string;
export const bubble: string;
export const bubbleActive: string;
export const add: string;
export const addActive: string;
