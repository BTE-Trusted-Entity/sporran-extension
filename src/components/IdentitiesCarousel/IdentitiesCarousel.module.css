.container {
  position: relative;
  width: 100%;
}

.chevron {
  position: absolute;
  top: 2rem;
  width: 1rem;
  height: 2rem;
}

.left {
  composes: chevron;
  background: url('../images/chevron-left.svg') no-repeat;
  left: 6.25rem;
}

.right {
  composes: chevron;
  background: url('../images/chevron-right.svg') no-repeat;
  right: 6.25rem;
}

.bubbles {
  list-style: none;
  margin: 0;
  padding: 0;
}

.item {
  display: inline-block;
}

.bubble {
  display: inline-block;
  height: 0.5rem;
  width: 0.5rem;
  background: var(--color-grey-light);
  opacity: 0.5;
  border-radius: 50%;
  margin: 0 0.25rem;
}

.bubbleActive {
  composes: bubble;
  opacity: 1;
}

.add {
  composes: bubble;
  composes: buttonIcon from '../Button/Button.module.css';
  background: url('../images/plus-circle.svg') no-repeat center/contain;
  opacity: 0.35;
}

.addActive {
  composes: add;
  opacity: 1;
}
