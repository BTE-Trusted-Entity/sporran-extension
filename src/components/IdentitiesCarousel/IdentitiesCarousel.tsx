import { Link, NavLink, useLocation, useRouteMatch } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';
import { sortBy } from 'lodash-es';

import * as styles from './IdentitiesCarousel.module.css';

import {
  Identity,
  isNew,
  NEW,
  useIdentities,
} from '../../utilities/identities/identities';
import { IdentitySlide } from '../IdentitySlide/IdentitySlide';
import { IdentitySlideNew } from '../IdentitySlide/IdentitySlideNew';
import { generatePath } from '../../views/paths';
import { isExtensionPopup } from '../../utilities/isExtensionPopup/isExtensionPopup';

interface IdentityLinkProps {
  identity: Identity;
  identities: Identity[];
  direction: 'previous' | 'next';
  showAdd: boolean;
}

function IdentityLink({
  identity,
  identities,
  direction,
  showAdd,
}: IdentityLinkProps): JSX.Element {
  const t = browser.i18n.getMessage;
  const { path } = useRouteMatch();
  const { search } = useLocation();

  const { length } = identities;

  const isPrevious = direction === 'previous';
  const delta = isPrevious ? -1 : 1;
  const modifiedIndex = !isNew(identity)
    ? identities.indexOf(identity) + delta
    : isPrevious
    ? length - 1
    : 0;

  const isInRange = 0 <= modifiedIndex && modifiedIndex < length;
  const isNewIdentity = showAdd && !isInRange;

  const linkedIndex = (modifiedIndex + length) % length;
  const linkedIdentity = isNewIdentity ? NEW : identities[linkedIndex];
  const title = isNewIdentity
    ? t('component_IdentityLink_title_new')
    : linkedIdentity.name;

  return (
    <Link
      to={generatePath(`${path}${search}`, { address: linkedIdentity.address })}
      title={title}
      aria-label={title}
      className={isPrevious ? styles.left : styles.right}
      replace
    />
  );
}

const maxIdentityBubbles = 5;

interface IdentitiesBubblesProps {
  identities: Identity[];
  showAdd: boolean;
}

export function IdentitiesBubbles({
  identities,
  showAdd,
}: IdentitiesBubblesProps): JSX.Element | null {
  const t = browser.i18n.getMessage;
  const { path } = useRouteMatch();
  const { search } = useLocation();

  const singleLink = !showAdd && identities.length <= 1;
  const tooManyLinks = identities.length > maxIdentityBubbles;
  if (singleLink || tooManyLinks) {
    return null; // hide the component when it makes no sense
  }

  return (
    <ul className={styles.bubbles}>
      {identities.map(({ name, address }) => (
        <li className={styles.item} key={address}>
          <NavLink
            className={styles.bubble}
            activeClassName={styles.bubbleActive}
            to={generatePath(`${path}${search}`, { address: address })}
            aria-label={name}
            title={name}
          />
        </li>
      ))}
      {showAdd && (
        <li className={styles.item}>
          <NavLink
            className={styles.add}
            activeClassName={styles.addActive}
            to={generatePath(path, { address: NEW.address })}
            aria-label={t('component_IdentitiesCarousel_title_new')}
            title={t('component_IdentitiesCarousel_title_new')}
          />
        </li>
      )}
    </ul>
  );
}

interface Props {
  identity: Identity;
  options?: boolean;
}

export function IdentitiesCarousel({
  identity,
  options = false,
}: Props): JSX.Element | null {
  const identities = useIdentities().data;
  if (!identities) {
    return null; // storage data pending
  }

  const identitiesList = sortBy(Object.values(identities), 'index');
  const showAdd = isExtensionPopup();
  const showLinks = showAdd || identitiesList.length > 1;

  return (
    <div className={styles.container}>
      {isNew(identity) ? (
        <IdentitySlideNew />
      ) : (
        <IdentitySlide identity={identity} options={options} />
      )}

      {showLinks && (
        <IdentityLink
          direction="previous"
          identity={identity}
          identities={identitiesList}
          showAdd={showAdd}
        />
      )}

      {showLinks && (
        <IdentityLink
          direction="next"
          identity={identity}
          identities={identitiesList}
          showAdd={showAdd}
        />
      )}

      <IdentitiesBubbles identities={identitiesList} showAdd={showAdd} />
    </div>
  );
}
