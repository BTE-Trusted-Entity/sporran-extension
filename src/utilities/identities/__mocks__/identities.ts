export const decryptIdentity = jest.fn();

export const saveIdentity = jest.fn();
