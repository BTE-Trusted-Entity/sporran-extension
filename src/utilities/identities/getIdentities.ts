import { storage } from '../storage/storage';

import { IdentitiesMap } from './types';

export const IDENTITIES_KEY = 'identities';

export async function getIdentities(): Promise<IdentitiesMap> {
  const stored = (await storage.get(IDENTITIES_KEY))[IDENTITIES_KEY];
  if (stored) {
    return stored;
  }

  await storage.set({ [IDENTITIES_KEY]: {} });
  return {};
}
