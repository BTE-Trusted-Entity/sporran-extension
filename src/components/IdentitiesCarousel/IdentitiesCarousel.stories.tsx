import { Meta } from '@storybook/react';

import {
  identitiesMock,
  IdentitiesProviderMock,
  moreIdentitiesMock,
} from '../../utilities/identities/IdentitiesProvider.mock';

import { IdentitiesCarousel } from './IdentitiesCarousel';

export default {
  title: 'Components/IdentitiesCarousel',
  component: IdentitiesCarousel,
  decorators: [(story) => <div style={{ textAlign: 'center' }}>{story()}</div>],
} as Meta;

export function Template(): JSX.Element {
  return (
    <IdentitiesCarousel
      identity={
        identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
      }
    />
  );
}

export function NoBubbles(): JSX.Element {
  return (
    <IdentitiesProviderMock identities={moreIdentitiesMock}>
      <IdentitiesCarousel
        identity={
          identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
        }
      />
    </IdentitiesProviderMock>
  );
}
