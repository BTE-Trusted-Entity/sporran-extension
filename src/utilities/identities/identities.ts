import { useContext } from 'react';
import { mutate } from 'swr';
import {
  blake2AsU8a,
  keyExtractPath,
  keyFromPath,
  mnemonicToMiniSecret,
  sr25519PairFromSeed,
} from '@polkadot/util-crypto';
import {
  Did,
  DidDocument,
  DidResourceUri,
  DidUri,
  EncryptRequestData,
  IEncryptedMessage,
  KiltAddress,
  KiltEncryptionKeypair,
  KiltKeyringPair,
  Message,
  MessageBody,
  SignCallback,
  SignRequestData,
  Utils,
} from '@kiltprotocol/sdk-js';
import { map, max, memoize } from 'lodash-es';

import {
  loadEncrypted,
  saveEncrypted,
} from '../storageEncryption/storageEncryption';

import { getDidDocument, getDidEncryptionKey, parseDidUri } from '../did/did';
import { storage } from '../storage/storage';
import { useSwrDataOrThrow } from '../useSwrDataOrThrow/useSwrDataOrThrow';

import { IdentitiesContext, IdentitiesContextType } from './IdentitiesContext';
import { getIdentities, IDENTITIES_KEY } from './getIdentities';

import { Identity } from './types';

export { Identity, IdentitiesMap } from './types';

const CURRENT_IDENTITY_KEY = 'currentIdentity';

export const NEW: Identity = {
  address: 'NEW' as KiltAddress,
  did: '' as DidUri,
  name: '',
  index: -1,
};

export function isNew(identity: Identity): boolean {
  return identity === NEW;
}

export function useIdentities(): IdentitiesContextType {
  return useContext(IdentitiesContext);
}

async function getCurrentIdentity(): Promise<string | null> {
  const identities = await getIdentities();
  const stored = await storage.get(CURRENT_IDENTITY_KEY);
  const current = stored[CURRENT_IDENTITY_KEY];

  if (identities[current]) {
    return current;
  }

  const firstIdentity = Object.values(identities)[0];
  if (!firstIdentity) {
    return null;
  }

  await setCurrentIdentity(firstIdentity.address);

  return firstIdentity.address;
}

export async function setCurrentIdentity(address: string): Promise<void> {
  const oldAddress = await storage.get(CURRENT_IDENTITY_KEY);
  if (address === oldAddress[CURRENT_IDENTITY_KEY]) {
    return;
  }
  await storage.set({ [CURRENT_IDENTITY_KEY]: address });
  await mutate(['getCurrentIdentity', CURRENT_IDENTITY_KEY]);
}

export async function saveIdentity(identity: Identity): Promise<void> {
  const identities = await getIdentities();
  identities[identity.address] = identity;
  await storage.set({ [IDENTITIES_KEY]: identities });
  await mutate(IDENTITIES_KEY);
}

export async function removeIdentity(identity: Identity): Promise<void> {
  const identities = await getIdentities();
  delete identities[identity.address];

  await storage.set({ [IDENTITIES_KEY]: identities });
  await storage.remove(identity.address);

  const firstIdentity = Object.values(identities)[0];
  await setCurrentIdentity(firstIdentity?.address);

  await mutate(IDENTITIES_KEY);
}

interface IdentityDidCrypto {
  didDocument: DidDocument;
  keypair: KiltKeyringPair;
  authenticationKey: KiltKeyringPair;
  sign: SignCallback;
  encrypt: (
    messageBody: MessageBody,
    dAppDidDocument: DidDocument,
  ) => Promise<IEncryptedMessage>;
}

export function deriveAuthenticationKey(seed: Uint8Array) {
  const baseKey = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  return baseKey.derive('//did//0') as typeof baseKey;
}

export function deriveEncryptionKeyFromSeed(
  seed: Uint8Array,
): KiltEncryptionKeypair {
  const keypair = sr25519PairFromSeed(seed);
  const { path } = keyExtractPath('//did//keyAgreement//0');
  const { secretKey } = keyFromPath(keypair, path, 'sr25519');
  return Utils.Crypto.makeEncryptionKeypairFromSeed(blake2AsU8a(secretKey));
}

export function deriveAttestationKeyFromSeed(seed: Uint8Array) {
  const baseKey = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  return baseKey.derive('//did//assertion//0') as typeof baseKey;
}

function deriveEncryptionKeyLegacy(seed: Uint8Array) {
  const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const encryptionKeyringPair = keypair.derive('//did//keyAgreement//0');

  const secret = encryptionKeyringPair
    .encryptMessage(
      new Uint8Array(24).fill(0),
      new Uint8Array(24).fill(0),
      new Uint8Array(24).fill(0),
    )
    .slice(24); // first 24 bytes are the nonce

  return Utils.Crypto.makeEncryptionKeypairFromSeed(secret);
}

async function fixLightDidIssues(seed: Uint8Array) {
  const identities = await getIdentities();
  const { address } = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const identity = identities[address];

  if (!identity) {
    // could be the Alice identity
    return;
  }

  const parsed = identity.did && Did.parse(identity.did);
  if (parsed && parsed.type !== 'light') {
    return;
  }

  try {
    if (!identity.did) {
      // DID was deactivated, no action needed.
      return;
    }

    // If this light DID was created and stored using SDK@0.24.0 then its keys are serialized using base64,
    // resulting in an invalid URI, so resolving would throw an exception.
    const document = await getDidDocument(identity.did);

    // Another issue we see is the light DIDs without key agreement keys, need to regenerate them as well
    const encryptionKey = getDidEncryptionKey(document);

    // This public key also means the DID needs to be regenerated
    const troubleKey =
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e';
    if (Utils.Crypto.u8aToHex(encryptionKey.publicKey) === troubleKey) {
      throw new Error();
    }
  } catch {
    // We re-create the invalid DID from scratch and update its URI in the identity.
    const { uri } = getLightDidFromSeed(seed);
    await saveIdentity({ ...identity, did: uri });
  }
}

export async function getIdentityCryptoFromSeed(
  seed: Uint8Array,
  legacy?: boolean,
): Promise<IdentityDidCrypto> {
  await fixLightDidIssues(seed);

  const authenticationKey = deriveAuthenticationKey(seed);
  const encryptionKey = legacy
    ? deriveEncryptionKeyLegacy(seed)
    : deriveEncryptionKeyFromSeed(seed);

  const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const identities = await getIdentities();
  const did = getIdentityDid(identities[keypair.address]);

  const didDocument = await getDidDocument(did);

  async function encryptCallback({ data, peerPublicKey }: EncryptRequestData) {
    const { box, nonce } = Utils.Crypto.encryptAsymmetric(
      data,
      peerPublicKey,
      encryptionKey.secretKey,
    );
    const keyUri = `${didDocument.uri}${getDidEncryptionKey(didDocument).id}`;
    return {
      nonce,
      data: box,
      keyUri: keyUri as DidResourceUri,
    };
  }

  async function sign({ data, keyRelationship }: SignRequestData) {
    if (keyRelationship !== 'authentication') {
      throw new Error(
        'Only key relationship "authentication" is supported for signing',
      );
    }

    const signature = authenticationKey.sign(data, { withType: false });
    const keyUri =
      `${didDocument.uri}${didDocument.authentication[0].id}` as DidResourceUri;
    const keyType = authenticationKey.type;

    return {
      signature,
      keyUri,
      keyType,
    };
  }

  async function encrypt(
    messageBody: MessageBody,
    dAppDidDocument: DidDocument,
  ): Promise<IEncryptedMessage> {
    const message = Message.fromBody(
      messageBody,
      didDocument.uri,
      dAppDidDocument.uri,
    );

    return Message.encrypt(
      message,
      encryptCallback,
      `${dAppDidDocument.uri}${getDidEncryptionKey(dAppDidDocument).id}`,
    );
  }

  return {
    didDocument,
    keypair,
    authenticationKey,
    sign,
    encrypt,
  };
}

export async function encryptIdentity(
  backupPhrase: string,
  password: string,
): Promise<KiltAddress> {
  const seed = mnemonicToMiniSecret(backupPhrase);
  const { address } = Utils.Crypto.makeKeypairFromUri(backupPhrase, 'sr25519');
  await saveEncrypted(address, password, seed);
  return address;
}

export function getLightDidFromSeed(seed: Uint8Array): DidDocument {
  return Did.createLightDidDocument({
    authentication: [deriveAuthenticationKey(seed)],
    keyAgreement: [deriveEncryptionKeyFromSeed(seed)],
  });
}

async function getIdentityName(): Promise<{ name: string; index: number }> {
  const identities = await getIdentities();
  const largestIndex = max(map(identities, 'index')) || 0;

  const index = 1 + largestIndex;

  return { name: `KILT Identity ${index}`, index };
}

export async function createIdentity(
  backupPhrase: string,
  password: string,
): Promise<Identity> {
  const address = await encryptIdentity(backupPhrase, password);

  const seed = mnemonicToMiniSecret(backupPhrase);

  const { uri } = getLightDidFromSeed(seed);

  const { name, index } = await getIdentityName();

  const identity = { name, address, did: uri, index };
  await saveIdentity(identity);

  return identity;
}

export async function importIdentity(
  backupPhrase: string,
  password: string,
): Promise<Identity> {
  const address = await encryptIdentity(backupPhrase, password);

  const seed = mnemonicToMiniSecret(backupPhrase);

  const lightDidDocument = getLightDidFromSeed(seed);
  const lightDid = lightDidDocument.uri;

  const resolved = await Did.resolve(lightDid);
  const fullDid = resolved?.metadata?.canonicalId;

  const deactivated = resolved?.metadata?.deactivated;
  const did = deactivated ? undefined : fullDid || lightDid;
  const deletedDid = deactivated ? Did.getFullDidUri(lightDid) : undefined;

  const { name, index } = await getIdentityName();

  const identity = {
    name,
    address,
    index,
    ...(!deactivated && { did }),
    ...(deactivated && { deletedDid }),
  };
  await saveIdentity(identity);

  return identity;
}

export async function decryptIdentity(
  address: string,
  password: string,
): Promise<Uint8Array> {
  const seed = await loadEncrypted(address, password);
  return new Uint8Array(seed);
}

/** Ensure that local information about the DID type matches stored on blockchain
 * even if an error occurred while asynchronous update was running */
async function syncDidStateWithBlockchain(address: string | null | undefined) {
  if (!address) {
    return;
  }

  const identities = await getIdentities();
  const identity = identities[address];

  if (!identity.did) {
    // could be a legacy identity without DID
    return;
  }

  const { fullDid, type } = parseDidUri(identity.did);
  const wasOnChain = type === 'full';

  const resolved = await Did.resolve(identity.did);
  const isOnChain = wasOnChain
    ? Boolean(resolved && resolved.metadata && !resolved.metadata.deactivated)
    : Boolean(resolved && resolved.metadata && resolved.metadata.canonicalId);

  if (wasOnChain && !isOnChain) {
    await saveIdentity({
      ...identity,
      did: undefined,
      deletedDid: fullDid,
    });
  }
  if (!wasOnChain && isOnChain) {
    await saveIdentity({ ...identity, did: fullDid });
  }
}

/** Memoized function will run only once per identity while the popup is open, do not use from backend */
const noAwaitUpdateCachedDidStateOnce = memoize(syncDidStateWithBlockchain);

export function useCurrentIdentity(): string | null | undefined {
  const data = useSwrDataOrThrow(
    CURRENT_IDENTITY_KEY,
    getCurrentIdentity,
    'getCurrentIdentity',
  );

  noAwaitUpdateCachedDidStateOnce(data);

  return data;
}

export function getIdentityDid({ did }: Identity): DidUri {
  if (!did) {
    throw new Error('DID is deleted and unusable');
  }
  return did;
}
