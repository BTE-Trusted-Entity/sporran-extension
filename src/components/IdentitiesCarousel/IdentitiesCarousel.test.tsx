import { MemoryRouter, Route } from 'react-router-dom';

import {
  identitiesMock as fewIdentities,
  moreIdentitiesMock as moreIdentities,
  render,
} from '../../testing/testing';
import { paths } from '../../views/paths';

import { NEW } from '../../utilities/identities/identities';

import { IdentitiesCarousel, IdentitiesBubbles } from './IdentitiesCarousel';

describe('IdentitiesCarousel', () => {
  it('should render normal identities', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.overview]}>
        <Route path={paths.identity.overview}>
          <IdentitiesCarousel
            identity={
              fewIdentities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
            }
          />
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the first identity', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.overview]}>
        <Route path={paths.identity.overview}>
          <IdentitiesCarousel
            identity={
              fewIdentities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
            }
          />
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the last identity', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.overview]}>
        <Route path={paths.identity.overview}>
          <IdentitiesCarousel
            identity={
              fewIdentities['4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653']
            }
          />
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the new identity', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.overview]}>
        <Route path={paths.identity.overview}>
          <IdentitiesCarousel identity={NEW} />
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should support other paths', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.send.start]}>
        <Route path={paths.identity.send.start}>
          <IdentitiesCarousel
            identity={
              fewIdentities['4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653']
            }
          />
        </Route>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render a bubble for each identity', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.overview]}>
        <Route path={paths.identity.overview}>
          <IdentitiesBubbles
            identities={Object.values(fewIdentities)}
            showAdd={true}
          />
        </Route>
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should not render bubbles if number of identities is more than the maximum', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={[paths.identity.overview]}>
        <Route path={paths.identity.overview}>
          <IdentitiesBubbles
            identities={Object.values(moreIdentities)}
            showAdd={true}
          />
        </Route>
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });
});
