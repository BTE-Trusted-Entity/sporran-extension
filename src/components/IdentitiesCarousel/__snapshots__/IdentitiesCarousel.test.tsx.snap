// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentitiesCarousel should not render bubbles if number of identities is more than the maximum 1`] = `<div />`;

exports[`IdentitiesCarousel should render a bubble for each identity 1`] = `
<div>
  <ul
    class="bubbles"
  >
    <li
      class="item"
    >
      <a
        aria-label="Light DID Identity"
        class="bubble"
        href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
        title="Light DID Identity"
      />
    </li>
    <li
      class="item"
    >
      <a
        aria-label="Identity with on-chain DID and lots of coins"
        class="bubble"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
        title="Identity with on-chain DID and lots of coins"
      />
    </li>
    <li
      class="item"
    >
      <a
        aria-label="Identity which has a very very very very long name and might not even fit on the screen"
        class="bubble"
        href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
        title="Identity which has a very very very very long name and might not even fit on the screen"
      />
    </li>
    <li
      class="item"
    >
      <a
        aria-label="Add Identity"
        class="add"
        href="/identity/NEW"
        title="Add Identity"
      />
    </li>
  </ul>
</div>
`;

exports[`IdentitiesCarousel should render normal identities 1`] = `
<div>
  <div
    class="container"
  >
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <a
      aria-label="Light DID Identity"
      class="left"
      href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
      title="Light DID Identity"
    />
    <a
      aria-label="Identity which has a very very very very long name and might not even fit on the screen"
      class="right"
      href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
      title="Identity which has a very very very very long name and might not even fit on the screen"
    />
    <ul
      class="bubbles"
    >
      <li
        class="item"
      >
        <a
          aria-label="Light DID Identity"
          class="bubble"
          href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
          title="Light DID Identity"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity with on-chain DID and lots of coins"
          class="bubble"
          href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
          title="Identity with on-chain DID and lots of coins"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity which has a very very very very long name and might not even fit on the screen"
          class="bubble"
          href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
          title="Identity which has a very very very very long name and might not even fit on the screen"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Add Identity"
          class="add"
          href="/identity/NEW"
          title="Add Identity"
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`IdentitiesCarousel should render the first identity 1`] = `
<div>
  <div
    class="container"
  >
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Light DID Identity
        </span>
      </div>
    </section>
    <a
      aria-label="Add Identity"
      class="left"
      href="/identity/NEW"
      title="Add Identity"
    />
    <a
      aria-label="Identity with on-chain DID and lots of coins"
      class="right"
      href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
      title="Identity with on-chain DID and lots of coins"
    />
    <ul
      class="bubbles"
    >
      <li
        class="item"
      >
        <a
          aria-label="Light DID Identity"
          class="bubble"
          href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
          title="Light DID Identity"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity with on-chain DID and lots of coins"
          class="bubble"
          href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
          title="Identity with on-chain DID and lots of coins"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity which has a very very very very long name and might not even fit on the screen"
          class="bubble"
          href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
          title="Identity which has a very very very very long name and might not even fit on the screen"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Add Identity"
          class="add"
          href="/identity/NEW"
          title="Add Identity"
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`IdentitiesCarousel should render the last identity 1`] = `
<div>
  <div
    class="container"
  >
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity which has a very very very very long name and might not even fit on the screen
        </span>
      </div>
    </section>
    <a
      aria-label="Identity with on-chain DID and lots of coins"
      class="left"
      href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
      title="Identity with on-chain DID and lots of coins"
    />
    <a
      aria-label="Add Identity"
      class="right"
      href="/identity/NEW"
      title="Add Identity"
    />
    <ul
      class="bubbles"
    >
      <li
        class="item"
      >
        <a
          aria-label="Light DID Identity"
          class="bubble"
          href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
          title="Light DID Identity"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity with on-chain DID and lots of coins"
          class="bubble"
          href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
          title="Identity with on-chain DID and lots of coins"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity which has a very very very very long name and might not even fit on the screen"
          class="bubble"
          href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
          title="Identity which has a very very very very long name and might not even fit on the screen"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Add Identity"
          class="add"
          href="/identity/NEW"
          title="Add Identity"
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`IdentitiesCarousel should render the new identity 1`] = `
<div>
  <div
    class="container"
  >
    <section>
      <a
        aria-label="Add Identity"
        class="add"
        href="/identity/add"
        title="Add Identity"
      />
    </section>
    <a
      aria-label="Identity which has a very very very very long name and might not even fit on the screen"
      class="left"
      href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
      title="Identity which has a very very very very long name and might not even fit on the screen"
    />
    <a
      aria-label="Light DID Identity"
      class="right"
      href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
      title="Light DID Identity"
    />
    <ul
      class="bubbles"
    >
      <li
        class="item"
      >
        <a
          aria-label="Light DID Identity"
          class="bubble"
          href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
          title="Light DID Identity"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity with on-chain DID and lots of coins"
          class="bubble"
          href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
          title="Identity with on-chain DID and lots of coins"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity which has a very very very very long name and might not even fit on the screen"
          class="bubble"
          href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
          title="Identity which has a very very very very long name and might not even fit on the screen"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Add Identity"
          class="add"
          href="/identity/NEW"
          title="Add Identity"
        />
      </li>
    </ul>
  </div>
</div>
`;

exports[`IdentitiesCarousel should support other paths 1`] = `
<div>
  <div
    class="container"
  >
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity which has a very very very very long name and might not even fit on the screen
        </span>
      </div>
    </section>
    <a
      aria-label="Identity with on-chain DID and lots of coins"
      class="left"
      href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/send"
      title="Identity with on-chain DID and lots of coins"
    />
    <a
      aria-label="Add Identity"
      class="right"
      href="/identity/NEW/send"
      title="Add Identity"
    />
    <ul
      class="bubbles"
    >
      <li
        class="item"
      >
        <a
          aria-label="Light DID Identity"
          class="bubble"
          href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/send"
          title="Light DID Identity"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity with on-chain DID and lots of coins"
          class="bubble"
          href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/send"
          title="Identity with on-chain DID and lots of coins"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Identity which has a very very very very long name and might not even fit on the screen"
          class="bubble"
          href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653/send"
          title="Identity which has a very very very very long name and might not even fit on the screen"
        />
      </li>
      <li
        class="item"
      >
        <a
          aria-label="Add Identity"
          class="add"
          href="/identity/NEW/send"
          title="Add Identity"
        />
      </li>
    </ul>
  </div>
</div>
`;
